### ISSUE TITLE
how to integrate optuna on custom-made objective function with certain HPs

### ISSUE THREAD

**Issue by ashraf-01** [optuna/optuna#1640]
HI all

my function attached in screen shot , under name "train'' with certain HPs mainly M , and learning rate , one float and the other is float , is there a way to integrate this in to the optuna framwork such as , that training can actually happen while optimizing for the HPs mentioned above , I Have tried this version included within the attachment , yet I get the error as showed within 

you notice that I have used 

    return theta, sharpes

but it must be the case that I want the optimized lb , lr  , so shall I use 

 return lb , lr ?

the logic is contained in actually in train() as in here 

def train(x , epochs = 2500, M=24, commission = 0.00002 , learning_rate = 0.3):  
     theta = np.random.rand(M + 2)
     sharpes = np.zeros(epochs)
     for i in range(epochs):
         grad, sharpe = gradient(x, theta, commission)
         theta = theta + grad * learning_rate 
         sharpes[i] = sharpe

    print("finished training")
    return theta, sharpes

![Screenshot from 2020-08-14 23-35-17](https://user-images.githubusercontent.com/50061659/90294800-9b8ece00-de87-11ea-9d0d-0c95f8c56041.png)Our input is highly appreciated 

![Screenshot from 2020-08-14 23-50-12](https://user-images.githubusercontent.com/50061659/90295381-1dcbc200-de89-11ea-9946-89d4718afd0c.png)

your input is highly appreciated

**Comment by HideakiImamura** [comment]
Thank you for the question! In Optuna standard learning scheme, the returned type of the objective function should be `float`. It is the value to be minimized such as the model's validation error. (Optuna supports maximization instead of minimization. So, it can be an accuracy instead of an error.)  I would like to know what do you want to minimize.

On the other hand, Optuna provides "multi-objective" optimization. We can optimize (minimize) the validation error and the execution time simultaneously. If you want to use the multi-objective optimization, please check [examples](https://github.com/optuna/optuna/tree/master/examples/multi_objective) or the [API documentation](https://optuna.readthedocs.io/en/stable/reference/multi_objective/index.html).

**Comment by ashraf-01** [comment]
Hi Hideaki ,

thanks for showing interest ,

 2 , 3 parameters of interest to this specific application are M (LOOKBACK PERIOD , INTEGER ) , NUMBER OF EPOCHS (LESS IMPORTANT ) AS IT IS ALREADY ALMOST OPTIMIZED , BUT  STILL NUMBER OF EPOCHS WILL CHANGE WITH THE CHANGE OF TIME FREQUENCY WHICH USED TO SAMPLE THE OBSERVATIONS , AND LASTLY THE LEARNING RATE (FLOAT ) , WOULD YOU LIKE TO SHARE WITH YOU DRAFT VERSION OF THE WHOLE APPLICATION ?

BEST REGARDS
________________________________

**Comment by HideakiImamura** [comment]
So, I'm not sure whether I've understood what you intend, but I would like to know what "objective function" do you want to minimize. The M (lookback period, integer), the number of epochs, and the learning rate are "input" of the objective function. 

Please check [our blog](https://medium.com/optuna/optuna-v1-86192cd09be5). This tells us the problem definition of hyperparameter tuning.

**Comment by ashraf-01** [comment]
here is a copy of the sequence of the steps , so that it is much better understood

looking forward to your response

Best Regards
________________________________

**Comment by ashraf-01** [comment]
sorry toshihikoyanase for teh late response , I thought it was already attached , anyway , I took a peek in to your email yesterday 
here is the link

**Comment by ashraf-01** [comment]
here is the full code 
def sharpe(rets):
    return rets.median() / rets.mad()


# In[100]:


# this is how positions are getting generated 
def positions(x, theta):   
    M = len(theta) - 2
    T = len(x)
    Ft = np.zeros(T)
    for t in range(M, T):
        xt = np.concatenate([[1], x[t - M:t], [Ft[t - 1]]])
        Ft[t] = np.tanh(np.dot(theta, xt))
    return Ft


# In[101]:


def returns(Ft, x, delta):
    T = len(x)
    rets = Ft[0:T - 1] * x[1:T] - delta * np.abs(Ft[1:T] - Ft[0:T - 1])
    return np.concatenate([[0], rets])

def gradient(x, theta, delta):   # we then compute the derivative of the Sharpe ratio with respect to theta
    Ft = positions(x, theta) 
    R = returns(Ft, x, delta)
    T = len(x)
    M = len(theta) - 2
    
    A = np.mean(R)
    B = np.mean(np.square(R))
    S = A / np.sqrt(B - A ** 2)

    dSdA = S * (1 + S ** 2) / A
    dSdB = -S ** 3 / 2 / A ** 2
    dAdR = 1. / T
    dBdR = 2. / T * R
    
    grad = np.zeros(M + 2)  # initialize gradient
    dFpdtheta = np.zeros(M + 2)  # for storing previous dFdtheta
    
    for t in range(M, T):
        xt = np.concatenate([[1], x[t - M:t], [Ft[t-1]]])
        dRdF = -delta * np.sign(Ft[t] - Ft[t-1])
        dRdFp = x[t] + delta * np.sign(Ft[t] - Ft[t-1])
        dFdtheta = (1 - Ft[t] ** 2) * (xt + theta[-1] * dFpdtheta)
        dSdtheta = (dSdA * dAdR + dSdB * dBdR[t]) * (dRdF * dFdtheta + dRdFp * dFpdtheta)
        grad = grad + dSdtheta
        dFpdtheta = dFdtheta
        
    return grad, S
def train(x , epochs = 2500, M=24, commission = 0.00002 , learning_rate = 0.3):  # try different values of the parameters
    theta = np.random.rand(M + 2)
    sharpes = np.zeros(epochs) # store sharpes over time
    for i in range(epochs):
        grad, sharpe = gradient(x, theta, commission)
        theta = theta + grad * learning_rate # THIS IS WHERE YOU UPDATE YOUR LEARNING RATE AND THETA 

        sharpes[i] = sharpe
    print("finished training")
    return theta, sharpes

x = np.array(rets)
    
N = 2000 # number of data points that will be used for training 
P = 519  # number of data points used for testing the agent 

x_train = x[-(N+P):-P]
x_test = x[-P:]

std = np.std(x_train)
mean = np.mean(x_train)

x_train = (x_train - mean) / std
x_test = (x_test - mean) / std

and here we should actually insert OPTUNA LAYER i believe 
np.random.seed(0)
theta, sharpes = train(x_train, epochs = 2500 , M = 24 , commission = 0.00002, learning_rate=0.3)
where epochs , M should be always between 1 and number of training data points , and finally learning rate 
your input is highly appreciated

**Comment by HideakiImamura** [comment]
Thank you for sharing the information. What problem do you solve? Is it regression or classification? Or, is it a reinforcement learning problem? Your code misses the `y` values against `x`.

**Comment by ashraf-01** [comment]
hi , just pointing to your question above it is an RL set up , so there is no classical y versus x approach as in classification or regression , yet x represents the sole input to this problem , usually you can add more , but just to give you some context , there is some implicit assumption made which is the tanh function above , it could be other ones , but we assume that these weights (thetas) are generated actually via tanh and what is the role of sharpe is to see if we can improve over tanh via gradient ascent on perfromance measure like sharpe or other , what I have added above is some modified version of the sharpe 
anyway , let me know if I can help more

**Comment by HideakiImamura** [comment]
Thanks for the explanation. First of all, I have to explain how optuna gets the optimized hyperparameter. Optuna assumes that there is an objective function whose inputs are hyperparameters and output is a **real value**. You have to create a real-valued function to be optimized when optimizing hyperparameters. Could you define some score using `theta` and `sharpes` in your setting? If you can, you can use Optuna to optimize `theta` and `sharpes`.

**Comment by ashraf-01** [comment]
I DID NOT GET YOUR POINT , AS SHARPES AND THETAS ARE THE QUANTITIES / PARAMETERS OF THE MODEL AND THESE ARE ALREADY OPTIMIZED USING THE gradient function , are you getting my point? but on the other hand the other inputs in train function , except for commission , are already HP ? and I wanna optimize for those and the actual parameters of the model , as far as I remember in the first screenshot above(hp_optimize function) , where it is showing that it is running , it shows as well the following "as Trialstate as Fail  "  BUT i have tried the other version where I included optuna layer inside the 
objective function but again it came up as before  "as Trialstate as Fail  "  here is again the screenshot 
![Screenshot from 2020-08-14 23-50-12](https://user-images.githubusercontent.com/50061659/91206841-1e295000-e708-11ea-9abf-39b505e5d753.png)
let me know if you fathom anything extra out of this , and thanks for the feedback

**Comment by HideakiImamura** [comment]
Your objective function has
- inputs: (`lb`, `lr`), and
- outputs: (`theta`, `sharpes`).
The `theta` and `sharpes` seem to be the `numpy.ndarray`. The pair of (`theta`, `sharpes`) does not have a type of `float` (real value) but `Tuple[ndarray, ndarray]`. You have to define the real-valued objective function to use any hyperparameter tuning framework. I think you have two options.

- [1]: Define a new real-valued objective function, which represents the "reward" of your problem setting.
- [2]: Use "multi-objective optimization" to optimize `theta` and `sharpes` simultaneously. In Optuna, you can use `optuna.multi_objective` modules. Please refer to the [examples](https://github.com/optuna/optuna/tree/master/examples/multi_objective) and the [API documentation](https://optuna.readthedocs.io/en/stable/reference/multi_objective/index.html). However, the number of objectives (` = len(theta) + len(sharpes)`) is too big according to the screenshot. I think the option [1] leads to better performance.

**Comment by toshihikoyanase** [comment]
> The `theta` and `sharpes` seem to be the `numpy.ndarray`.

I agree with @HideakiImamura . The objective function is supposed to return a `float` value instead of `numpy.ndarray`.

You may already know, but I think the [first optimization of the tutorial](https://optuna.readthedocs.io/en/stable/tutorial/first.html) demonstrates what kind of the objective function Optuna can handle.

**Comment by ashraf-01** [comment]
I will let you know , if I got it right , thanks for the suggestions
